강원대 BE_오승규 2주차 과제(1단계)

5win opened the pull request:
안녕하세요 멘토님 매번 피드백 감사드립니다!
아직 반영하지 못했었던 피드백과 몇 가지 수정 사항을 반영하면서 step1을 마쳤습니다.
그 과정에서 고민되는 부분이나 의견을 여쭙고 싶은 부분을 몇 가지 남기고자 합니다 🙂


## 1. 반환할 오류 상태 코드
입력에 대해 유효성을 검증한 뒤, 통과하지 못하면 `404 Not Found` 상태코드와 함께 에러 메시지를 반환하도록 구현했습니다.
여러 상태코드들을 보면서 무엇을 반환할지 고민하긴 했으나, 마땅한 코드가 없었다고 생각합니다.

때문에 좀 더 구체적인 뜻을 나타내는 `4xx` 코드를 보내기 보다는 포괄적으로 자주 쓰이는 `404`를 반환하도록 했습니다.
혹시 더 적합하다고 생각하시는 상태코드나 처리 로직이 생각나시면 조언해주세요!

## 2. 2주차 과제의 요구사항 중 담당 MD와의 협의에 대한 내용
현재 해당 요구사항에 대해서 "카카오"라는 단어가 있을 경우 에러 메시지로 "담당 MD와 협의 후 사용할 수 있습니다" 와 같이 메시지를 보내주는 정도로 구현했습니다.

여기서 궁금한 점이 몇가지 있었습니다.
1.  관리자 페이지인데 관리자가 사용할 수 없고 MD를 거쳐야하는 경우가 따로 있을수도 있을까요?
2.  추후에 MD와 협의가 완료됐을 경우, 이 사실을 어떻게 알고 추가하는 것이 좋을까요?

두 번째 질문에 대한 제 생각은
1. 관리자 아이디에게 어떠한 권한?을 부여해준다
    1. 이 경우에는 포괄적인 권한이 아닌, 해당 "상품명"에 대한 권한만 부여해야 할 것 같습니다.
    2. (이러한 경우는 어떤 방식으로 구현할 수 있는지 아직 잘 모르겠습니다...)
2. MD가 직접 추가할 수 있도록 하며, MD가 추가할 수 있도록 엔드포인트를 추가적으로 만들어야 한다.
    1. 또는, MD의 계정에는 추가적인 권한이 있어서, 기존 엔드포인트로 얼마든지 "카카오"가 있는 상품명을 추가할 수 있게 한다.
       
이렇게 두 가지 정도가 생각나는데 일반적으로 어떻게 로직을 구성하는지 알고 싶습니다!

## 3. `ExceptionHandler`에 대한 테스트
유효성 검증 시에 발생하는 오류를 `GlobalExceptionHandler.java`에서 처리하도록 했습니다.
그리고 `ProductController`의 테스트 코드에서 유효성 검증에 대한 테스트를 진행하도록 구성했는데요
이 때 `GlobalExceptionHandler.java`의 에러 핸들링에 대한 테스트가 함께 진행된다고 생각하여 따로 테스트 코드를 작성하진 않았습니다.

이렇게 다른 테스트 코드에서 로직이 테스트 되는 경우에도 따로 테스트를 진행해야 할까요?
혹시 작성한다면 `MethodArgumentNotValidException` 객체를 직접 던져줘야 할 것 같은데 이런 식으로 구현하면 될지도 조언 부탁드려요!

## 3. `GET` 요청에 대한 예외 처리
`POST`, `PUT`, `DELETE`에 대해서는 `404 Not Found` 상태를 반환하여 메시지를 출력하도록 했습니다.
하지만 아직 존재하지 않는 `id`에 대한 `GET /api/products/product/{id}` 요청의 예외처리를 하지 않아 whitelabel 페이지가 출력됩니다.

`GET` 엔드포인트들은 `view`를 반환하게 되어있는데,
이런 경우에는 그냥 whitelabel을 그대로 두는게 나을까요?, 아니면 따로 `404` 페이지를 구성해서 `view`로 반환하는게 나을까요?


## p.s.
몇 가지 리뷰 받고 싶은 부분을 코드에 코멘트?로 남겨두었는데 이런식으로 리뷰를 받아도 되는지 잘 모르겠네요...ㅎ
그리고 혹시 추가적인 질문을 드리고 싶을 때는 slack DM으로 드려도 되는지 여쭤보고 싶습니다!

5win committed:
feat: Copy week1 assignment's code
- step3 제출한 코드에서 조금 더 보완한 코드를 사용.

5win committed:
refact: Changed to use @RequestBody instead of @Requestparam
- ProductController.java
	- JSON 데이터를 받아 @RequestBody를 통해 객체로 바인딩 되도록 변경.
- httpReq.js
	- 데이터를 JSON으로 변환하는 작업 추가.
	- HTTP Content-Type을 application/json으로 변경.

5win committed:
test: Changed test code to use JSON content type and PUT, DELETE method
- JSON 타입으로 입력을 변경.
- 컨트롤러의 PUT, DELETE로의 변경에 따라 테스트 코드도 수정.
- 컨트롤러의 응답 코드의 변경에 따라 비교할 status값 변경.

5win committed:
refact: Added productRowMapper member variable
- selectProducts()와 selectOneProduct의 RowMapper 람다식 중복 제거.

5win committed:
feat: Added input validation functions
- 상품명은 공백 포함 최대 15자까지만 입력 가능.
- 상품명에는 한글, 영어, 숫자, 일부 특수 문자만 입력 가능.
    - 특수 문자는 (,),[,],+,-,&,/,_ 만 허용.
- 상품명에 "카카오" 문구를 포함하려면 담당 MD와 협의 후 사용 가능.
- 유효성 검증 실패 시, 404 Not Found와 함께 메시지 반환.

5win committed:
refact: Changed endpoints URI
- GET /api/products/add -> /api/products/product
- POST /api/products/add -> /api/products/product
- GET /api/products/{id} -> /api/products/product/{id}
- PUT /api/products/edit -> /api/products/product
- DELETE /api/products/{id} -> /api/products/product/{id}

5win committed:
test: Added input name validation test code
- 성공 케이스, 실패 케이스 모두 추가함.

5win committed:
refact: Change duplicated or not exists id validation logic
- ProductDao.selectOneProduct
	- id에 해당하는 객체 반환에서 객체 리스트를 반환하도록 수정.
	- 정상적인 상태라면 1개의 객체를 담아서 반환.
- ProductController.java
	- selectOneProduct()를 통해 id 중복, 존재 여부 확인.
	- 유효하지 않으면 404 Not Found.
- 변경 로직에 맞게 테스트 코드 수정.

5win committed:
refact: changed HTTP 200 OK response body
- 200 OK 응답 전송 시, 작업 성공 메시지를 반환하도록 변경.
- 리다이렉션 URI는 ajax에 명시하도록 변경.

5win committed:
Merge branch '5win' into step1

5win committed:
docs: Update README.md
- 2주차 step1의 요구사항에 맞게 업데이트.

5win committed:
test: Added error message test code
- validation과 관련된 오류 메시지 테스트.
- id 중복 또는 존재하지 않을 때의 오류 메시지 테스트.

5win committed:
docs: Fix README.md

5win commented on src/main/java/gift/Product.java:
```diff
@@ -1,5 +1,27 @@
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Pattern;
+import jakarta.validation.constraints.Size;
+
```
어떻게 작성해야 스타일을 지키면서 가독성을 높일지 고민했으나, 어떤게 좋은지 확신이 서지 않네요..

메시지 문자열이나 정규표현식 문자열을 따로 상수로 빼내면 더 좋을 것 같은데 `record`라서 해당 클래스 내에서는 안될 것 같았습니다.
`enum`과 같은 것을 구현해서 이러한 메시지들을 따로 관리하는 것이 더 좋을까요?

jaymarystella replied:
그럼 class를 활용하면 어떨까요?
지금 상황에서는 메세지는 꼭 enum으로 관리하지 않으셔도됩니다.

5win replied:
> 그럼 class를 활용하면 어떨까요? 지금 상황에서는 메세지는 꼭 enum으로 관리하지 않으셔도됩니다.

넵 반영하여 상수 클래스들을 좀 분리해봤습니다!
혹시 이러한 상황에선 왜 enum보다는 class가 낫다고 생각하시는지 여쭤봐도 될까요?

jaymarystella replied:
지금처럼 몇개 없는 간단한 경우는 클래스로 만드는것이 더 간편한데요, 당연히 안되는건아니고 enum도 그룹화나 타입안정성측면에서 장점이 있어서 사용의사가있으시면 사용해보셔도됩니다.

5win commented on src/main/java/gift/ProductController.java:
```diff
@@ -40,28 +41,23 @@ public String getProducts(Model model) {
-            return ResponseEntity.badRequest().body("이미 존재하는 id입니다.");
+    @PostMapping("/product")
+    public ResponseEntity<String> addProduct(@RequestBody @Valid Product product) {
+        if (!productDao.selectOneProduct(product.id()).isEmpty()) {
```
멘토님의 피드백을 반영하여, `id`가 중복되거나 없는 상품일 경우, 예외처리를 하지 않고 쿼리를 통해 미리 확인하는 방식으로 바꿔보았습니다.

`selectOneProduct` 메서드를 보시면 아시겠지만 기존엔 `single()`을 통해 하나의 객체를 반환했었습니다.
그러나 `id`에 해당하는 상품이 없을 경우에는 `null`을 반환하지 않고 에러를 발생시키더라구요..

이에 대한 해결로 두 가지를 생각했습니다.

1. `selectOneProduct()`에서 에러를 캐치하고 대신 `null`을 반환한다. 따라서 `null`이 아닐 경우에는 중복인 상황이다.
2. `selectProducts()` 처럼 `List`를 반환하게 한 뒤, `size`가 1이상이면 중복인 상황이다.

이왕이면 NPE를 피하도록 하는 것이 좋다고 들어서 `null`을 반환하는 것은 지양하고 `List`를 반환하도록 결정했습니다.
괜찮은 선택이었는지, 더 좋은 방법이 있는지 피드백 부탁드립니다!

jaymarystella replied:
1. null반환은 지양해주세요. optional로 처리해보시면 좋을듯합니다.
2. 응답타입이 적절치 않습니다. 사용처에서 get(0)로 접근하여 빼야하기도하고 코드 가독성이 매우 나빠질거같아요. 메서드명이나 반환타입을보고 행동을 유추할 수 있어야합니다.

5win replied:
> 1. null반환은 지양해주세요. optional로 처리해보시면 좋을듯합니다.
> 2. 응답타입이 적절치 않습니다. 사용처에서 get(0)로 접근하여 빼야하기도하고 코드 가독성이 매우 나빠질거같아요. 메서드명이나 반환타입을보고 행동을 유추할 수 있어야합니다.

찾아보니 single이 아닌 optional로 바로 반환이 되더라구요...ㅎㅎ
반영하여 optional로 처리하도록 수정했습니다!
그리고 추가적으로 에러를 던지도록 바꾼 뒤, GlobalExceptionHandler에서 처리하도록 수정해봤습니다.

5win commented on src/main/java/gift/ProductController.java:
```diff
@@ -40,28 +41,23 @@ public String getProducts(Model model) {
-        } catch (DataAccessException e) {
-            return ResponseEntity.badRequest().body("이미 존재하는 id입니다.");
+    @PostMapping("/product")
+    public ResponseEntity<String> addProduct(@RequestBody @Valid Product product) {
```
이전에는 HTTP 응답의 개념에 대해서도 헷갈렸고 form에서 JSON방식과 다른 방식의 파라미터를 받는 방법도 어려워서 `@RequestParam`으로 받도록 했으나, 이번에 JSON 형식으로 타입으로 바꾸고 `@ResponseBody`를 사용하도록 변경해보았습니다.

jaymarystella commented on src/main/java/gift/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+     *
+     * @return 404 Not Found 상태코드와 함께 메시지를 반환
+     */
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<String> inputValidException(MethodArgumentNotValidException e) {
+        String errorMsg = Objects.requireNonNull(e.getBindingResult().getFieldError(),
+            "에러 필드가 없습니다.").getDefaultMessage();
+
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(errorMsg);
```
> 입력에 대해 유효성을 검증한 뒤, 통과하지 못하면 404 Not Found 상태코드와 함께 에러 메시지를 반환하도록 구현했습니다.
> 여러 상태코드들을 보면서 무엇을 반환할지 고민하긴 했으나, 마땅한 코드가 없었다고 생각합니다.
> 
> 때문에 좀 더 구체적인 뜻을 나타내는 4xx 코드를 보내기 보다는 포괄적으로 자주 쓰이는 404를 반환하도록 했습니다.
> 혹시 더 적합하다고 생각하시는 상태코드나 처리 로직이 생각나시면 조언해주세요!

NOT_FOUND는 리소스를 찾을 수 없을때 반환합니다. 하지만 handling하는 예외를 보면 요청인자가 잘못되었을때를 handle하시려는 의도로 보입니다. BAD_REQUEST가 더 맞지 않을까요?

5win replied:
> > 입력에 대해 유효성을 검증한 뒤, 통과하지 못하면 404 Not Found 상태코드와 함께 에러 메시지를 반환하도록 구현했습니다.
> > 여러 상태코드들을 보면서 무엇을 반환할지 고민하긴 했으나, 마땅한 코드가 없었다고 생각합니다.
> > 때문에 좀 더 구체적인 뜻을 나타내는 4xx 코드를 보내기 보다는 포괄적으로 자주 쓰이는 404를 반환하도록 했습니다.
> > 혹시 더 적합하다고 생각하시는 상태코드나 처리 로직이 생각나시면 조언해주세요!
> 
> NOT_FOUND는 리소스를 찾을 수 없을때 반환합니다. 하지만 handling하는 예외를 보면 요청인자가 잘못되었을때를 handle하시려는 의도로 보입니다. BAD_REQUEST가 더 맞지 않을까요?

저도 BAD_REQUEST가 적합하다고 생각되어 수정했습니다 감사합니다!

jaymarystella commented on src/main/java/gift/Product.java:
```diff
@@ -1,5 +1,27 @@
+    @Pattern(
+        regexp = "^[a-zA-Z0-9가-힣\\s\\(\\)\\[\\]\\+\\-\\&\\/\\_]*$",
+        message = "상품명에는 특수 문자 (,),[,],+,-,&,/,_ 만 허용됩니다.")
+    @Pattern(
```
> 현재 해당 요구사항에 대해서 "카카오"라는 단어가 있을 경우 에러 메시지로 "담당 MD와 협의 후 사용할 수 있습니다" 와 같이 메시지를 보내주는 정도로 구현했습니다.
> 
> 여기서 궁금한 점이 몇가지 있었습니다.
> 
> 관리자 페이지인데 관리자가 사용할 수 없고 MD를 거쳐야하는 경우가 따로 있을수도 있을까요?
> 추후에 MD와 협의가 완료됐을 경우, 이 사실을 어떻게 알고 추가하는 것이 좋을까요?
> 두 번째 질문에 대한 제 생각은
> 
> 관리자 아이디에게 어떠한 권한?을 부여해준다
> 이 경우에는 포괄적인 권한이 아닌, 해당 "상품명"에 대한 권한만 부여해야 할 것 같습니다.
> (이러한 경우는 어떤 방식으로 구현할 수 있는지 아직 잘 모르겠습니다...)
> MD가 직접 추가할 수 있도록 하며, MD가 추가할 수 있도록 엔드포인트를 추가적으로 만들어야 한다.
> 또는, MD의 계정에는 추가적인 권한이 있어서, 기존 엔드포인트로 얼마든지 "카카오"가 있는 상품명을 추가할 수 있게 한다.
> 이렇게 두 가지 정도가 생각나는데 일반적으로 어떻게 로직을 구성하는지 알고 싶습니다!

1. 이거는 정책 나름일거같아요.
2. 이것도 구현나름일텐데, 제일간단하게 생각나는방법은, 요청이 올때마다 카카오라는 문구가 포함되어있다면 사전에 정의된 MD가 허용하는 "카카오"가 포함된 문구를 가진 테이블을 참조하는 방식으로 구현하면 될 것 같아요.

2번의 연장선으로 pattern으로 해주셨지만 내부적으로 애플리케이션 코드내에서 검증하도록 하는것도 추가가 되면 좋을거같아요.

말씀하신대로 계정별로 허용하는것도 방법이 될거구요. 정책을 얼마나 타이트하게 가져가냐, MD를 믿고 모든 권한을 위임하냐 의 차이로 보입니다.

5win replied:
> 
> 2번의 연장선으로 pattern으로 해주셨지만 내부적으로 애플리케이션 코드내에서 검증하도록 하는것도 추가가 되면 좋을거같아요.

지금은 `@Valid`에 의해 컨트롤러에서 파라미터로 받는 과정에서 검증이 되고 있는데, 
애플리케이션 코드 내라면 컨트롤러에서 파라미터를 받은 뒤에 엔드포인트 메서드 내에서 다시 검증하는 것을 말씀하시는 걸까요?

jaymarystella replied:
엔드포인트 메서드가 아니라 서비스클래스내에서 검증하는부분을 의미했습니다. 물론 valid에서 걸릴거긴해요. 다만 다른 클래스에서 사용될때 빠트린다면 구멍이 생길수있는데요, 서비스에서 한번더 검증을 해주는게 (혹은 도메인의 책임으로 본다면 도메인에 해당 역할을 가져가도 되겠구요) 안전하기때문입니다. 
지금에야 간단하지만 프로젝트가 점점 커지고 모든 시스템을 다 알 수 없는 상태가되면 이러한 실수는 비일비재하게 일어납니다.

jaymarystella commented on src/main/java/gift/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
> 유효성 검증 시에 발생하는 오류를 GlobalExceptionHandler.java에서 처리하도록 했습니다.
> 그리고 ProductController의 테스트 코드에서 유효성 검증에 대한 테스트를 진행하도록 구성했는데요
> 이 때 GlobalExceptionHandler.java의 에러 핸들링에 대한 테스트가 함께 진행된다고 생각하여 따로 테스트 코드를 작성하진 않았습니다.
> 
> 이렇게 다른 테스트 코드에서 로직이 테스트 되는 경우에도 따로 테스트를 진행해야 할까요?
> 혹시 작성한다면 MethodArgumentNotValidException 객체를 직접 던져줘야 할 것 같은데 이런 식으로 구현하면 될지도 조언 부탁드려요!

넵 말씀주신대로 컨트롤러 테스트에서 검증이 됐으니 굳이 따로 만들필요는 없을거같아요.

jaymarystella commented on src/main/java/gift/ProductController.java:
```diff
@@ -40,28 +41,23 @@ public String getProducts(Model model) {
      * @return addForm.html
      */
-    @GetMapping("/add")
+    @GetMapping("/product")
```
> 3. GET 요청에 대한 예외 처리
> POST, PUT, DELETE에 대해서는 404 Not Found 상태를 반환하여 메시지를 출력하도록 했습니다.
> 하지만 아직 존재하지 않는 id에 대한 GET /api/products/product/{id} 요청의 예외처리를 하지 않아 whitelabel 페이지가 출력됩니다.
> 
> GET 엔드포인트들은 view를 반환하게 되어있는데,
> 이런 경우에는 그냥 whitelabel을 그대로 두는게 나을까요?, 아니면 따로 404 페이지를 구성해서 view로 반환하는게 나을까요?

일단은 뷰가 붙은 애플리케이션이니 404페이지를 만드는것이 깔끔할것같습니다.

jaymarystella commented on src/main/java/gift/ProductController.java:
```diff
@@ -10,8 +11,8 @@
+import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestParam;
 
```
> 몇 가지 리뷰 받고 싶은 부분을 코드에 코멘트?로 남겨두었는데 이런식으로 리뷰를 받아도 되는지 잘 모르겠네요...ㅎ
> 그리고 혹시 추가적인 질문을 드리고 싶을 때는 slack DM으로 드려도 되는지 여쭤보고 싶습니다!

물론입니다. 질문은 어떤 통로로든 자유롭게 하셔도됩니다. 눈치보지말고 물어보고싶은것 다 물어보세요. 제가 답변드릴 수 있는 선에선 답변드리겠습니다.

jaymarystella commented on src/main/java/gift/ProductController.java:
```diff
@@ -70,44 +66,38 @@ public ResponseEntity<String> addProduct(
-            productDao.deleteProduct(id);
-        } catch (DataAccessException e) {
-            return ResponseEntity.badRequest().body("잘못된 요청입니다.");
+    @DeleteMapping("/product/{id}")
```
뷰가 필요한 controller와 그렇지 않은 controller를 분리해보시면 좋을거같아요.

5win replied:
> 뷰가 필요한 controller와 그렇지 않은 controller를 분리해보시면 좋을거같아요.

반영하여 뷰를 반환하는 GET 요청에 대한 엔드포인트들은 따로 분리해보았습니다!

jaymarystella commented on src/main/java/gift/Product.java:
```diff
@@ -1,5 +1,27 @@
 package gift;
 
-public record Product(long id, String name, long price, String imageUrl) {
+import jakarta.validation.constraints.NotBlank;
```
gift내 패키지구조를 어떻게 나누면 좋을까요?

5win replied:
> gift내 패키지구조를 어떻게 나누면 좋을까요?

반영하여 나눠보았는데 혹시 어색한 부분이 있다면 말씀해주세요!

jaymarystella replied:
패키지명에 대문자가있는거같네요

5win replied:
> 패키지명에 대문자가있는거같네요

수정하겠습니다!

jaymarystella reviewed (changes requested):
안녕하세요 승규님.
1단계 구현 고생많으셨습니다.
학습을 열심히 잘 해주셨네요.
질문주신부분 답변 남겨두었고 해결되지 않은 부분은 다시말씀주세요.

5win committed:
refact: Change duplicated or not exists id validation logic and HTTP status code.
- ProductDao.selecOneProduct
	- id에 해당하는 객체를 optional로 감싸서 반환.
- ProductController
	- 반환받은 optional 객체가 null이면 에러 throw
- GlobalExceptionHandler
	- 던져진 에러를 처리하는 핸들러 추가
	- NoSuchElementException 핸들러
	- IllegalArgumentException 핸들러
- 이에 맞게 테스트 코드도 수정.

5win committed:
refact: Seperate controller endpoints
- view를 반환하는 엔드포인트를 ProductViewController로 분리.

5win committed:
refact: Seperate package structures

5win committed:
refact: Add contants package and class
- 상수를 모아놓은 클래스를 정의하고 기존 하드코딩 상수들을 대체함.

5win committed:
docs: Update comments

5win commented:
@jaymarystella 
항상 리뷰 감사합니다 ☺️
step2를 진행하기 위해서 bearer 토큰을 공부했습니다.
그리고 적용해보고자 찾아보니 대부분이 spring security를 사용하고 있더라고요. 그래서 저도 문서도 보고 하면서 해보려고 했으나 생각보다 지금 프로젝트에 좀 과한? 느낌이 들정도로 거대한 로직인 것 같아서 고민하던 찰나에 강사님이 spring security 없이 구현하는 것을 의도했다고 하셔서 방향을 바꾸려고 합니다.

내일 18시까지 제출이라 시간이 많지 않아서 조언을 좀 해주시면 감사할 것 같습니다 ㅎㅎ
일단 제가 구상한 방식은 다음과 같습니다
1. 회원가입 서비스 구현하고 회원가입시에 member 테이블에 저장
2. 로그인 요청 시, 서버가 액세스 토큰을 만들어 해당 유저와 매핑?한 뒤 토큰을 응답으로 반환
3. 로그인된 유저는 매 요청마다 액세스 토큰을 함께 전달
4. 로그아웃 시, 서버는 해당 토큰을 삭제

이 경우에 로그아웃 시 토큰을 삭제하는게 맞을까요?
로그아웃을 안하고 일정시간이 지나면 다시 로그인하도록 하려면 유효기간을 정해야할 것 같은데 좀 포괄적이지만 조언 부탁드립니다..


그리고 위시리스트 접근에 대한 생각입니다.
1. 클라이언트가 위시리스트 접근 요청 시, 서버는 토큰을 검증
    1. 토큰이 없다면, forbidden 반환
    2. 토큰이 있으면 
        1. 해당 토큰에 매칭되는 유저의 위시리스트를 반환?
        2. 아니면 요청시에 username도 함께 보내주게 하고, username과 토큰을 통해 위시리스트를 반환?
2. 검증이 완료되면 리소스를 반환

---
추가)
힌트를 보고 jjwt를 살짝 보니까 claim이라는 것을 통해 이름, 유효기간, 토큰을 관리할 수 있는 것 같네요
이 부분은 더 공부해보겠습니다!

jaymarystella commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    @PostMapping()
+    public ResponseEntity<String> addProduct(@RequestBody @Valid Product product) {
+        productDao.selectOneProduct(product.id())
+            .ifPresent(v -> {
+                throw new IllegalArgumentException(ErrorMessage.ID_ALREADY_EXISTS_MSG);
+            });
+        productDao.insertNewProduct(product);
```
훨씬 깔끔해졌네요. 잘 만들어주셨습니다.

한걸음 더 나아가면 좋을거같습니다. 컨트롤러는 최대한 얇게 유지를 하는게 좋은데요, 어떻게 개선하면 좋을까요?

jaymarystella replied:
다음 단계때 고민해주시고 반영해주세요. 어려운점있으시면 말씀주시구요

jaymarystella commented:
과제플랫폼을 보니
`Base64로 인코딩된 사용자 ID, 비밀번호 쌍을 인증 정보(credentials) 값으로 사용한다.`
라고 되어있네요. 딱 요정도라면 비밀번호를 바꾸지 않는이상 항상 같은 값일거같은데요, 굳이 삭제를 안해도되겠네요. 

말씀주신대로 유효기간등을 고려해보는건 추가적인 사항이 될거같고 여력이 되면 구현해보는것도 재밌을것같습니다. 일단은 요구사항을 만족한뒤 개선해보시죠.

두번째 위시리스트 관련은 토큰으로 인증을 했으면 사용자가 식별이 될텐데 username을 따로 받는다는 부분이 잘 와닿지 않아서 위시리스트가 username으로 관리가 되는걸까요?

jaymarystella reviewed (changes requested):
안녕하세요 승규님.
2단계 잘 구현해주셨습니다.
구조적으로 한부분을 개선해주시면 좋을거같아요.
공부도 많이 해보셨네요. 수정 후 재요청부탁드려요.

jaymarystella commented:
아 다음단계때 반영주시면 될거같아요.

jaymarystella reviewed (approved):
안녕하세요 승규님.
2단계 잘 구현해주셨습니다.
구조적으로 한부분을 개선해주시면 좋을거같아요.
공부도 많이 해보셨네요. 수정 후 다음단계에 반영해주세요.

5win commented:
> 과제플랫폼을 보니 `Base64로 인코딩된 사용자 ID, 비밀번호 쌍을 인증 정보(credentials) 값으로 사용한다.` 라고 되어있네요. 딱 요정도라면 비밀번호를 바꾸지 않는이상 항상 같은 값일거같은데요, 굳이 삭제를 안해도되겠네요.
> 
> 말씀주신대로 유효기간등을 고려해보는건 추가적인 사항이 될거같고 여력이 되면 구현해보는것도 재밌을것같습니다. 일단은 요구사항을 만족한뒤 개선해보시죠.
> 
> 두번째 위시리스트 관련은 토큰으로 인증을 했으면 사용자가 식별이 될텐데 username을 따로 받는다는 부분이 잘 와닿지 않아서 위시리스트가 username으로 관리가 되는걸까요?

넵 일단 요구사항 구현을 우선적으로 해보겠습니다.
username부분은 제가 이해가 좀 부족했던 것 같습니다. 감사합니다!
